tests/testthat/test-eval-relocate.R

```diff
@@ -0,0 +1,192 @@
+    relocate_loc(x, y, after = where(is.character)),
+    c(x = 1L, z = 3L, y = 2L)
+  )
+})
+
+test_that("minimum of the `before` selection is used", {
+  x <- c(a = 1, b = 1, c = 1, d = 1, e = 1)
+  expect_named(relocate_loc(x, e, before = c(b, d)), c("a", "e", "b", "c", "d"))
+})
+
+test_that("maximum of the `after` selection is used", {
+  x <- c(a = 1, b = 1, c = 1, d = 1, e = 1)
+  expect_named(relocate_loc(x, b, after = c(a, c, e)), c("a", "c", "d", "e", "b"))
+})
+
+test_that("works with zero column data frames (tidyverse/dplyr#6167)", {
+  expect_identical(relocate_loc(data.frame(), any_of("b")), named(int()))
+  expect_identical(relocate_loc(data.frame(), any_of("b"), before = where(is.character)), named(int()))
+  expect_identical(relocate_loc(data.frame(), any_of("b"), after = where(is.character)), named(int()))
+})
+
+test_that("retains the last duplicate when renaming while moving (tidyverse/dplyr#6209)", {
+  # To enforce the invariant that relocating can't change the number of columns
+  x <- c(x = 1)
+  expect_identical(
+    relocate_loc(x, c(a = x, b = x)),
+    c(b = 1L)
+  )
+
+  x <- c(x = 1, y = 2)
+  expect_identical(
+    relocate_loc(x, c(a = x, b = y, c = x)),
+    c(b = 2L, c = 1L)
+  )
+})
+
+test_that("respects order specified by `...` (tidyverse/dplyr#5328)", {
+  x <- c(a = 1, x = 1, b = 1, z = 1, y = 1)
+
+  expect_named(
+    relocate_loc(x, c(x, y, z), before = x),
+    c("a", "x", "y", "z", "b")
+  )
+  expect_named(
+    relocate_loc(x, c(x, y, z), after = last_col()),
+    c("a", "b", "x", "y", "z")
+  )
+  expect_named(
+    relocate_loc(x, c(x, a, z)),
+    c("x", "a", "z", "b", "y")
+  )
+})
+
+test_that("allows for renaming (tidyverse/dplyr#5569)", {
+  x <- c(a = 1, b = 1, c = 1)
+
+  expect_named(
+    relocate_loc(x, c(new = b)),
+    c("new", "a", "c")
+  )
+  expect_named(
+    relocate_loc(x, c(new = b), after = c),
+    c("a", "c", "new")
+  )
+})
+
+test_that("can't supply both `before` and `after`", {
+  expect_snapshot(error = TRUE, {
+    relocate_loc(c(x = 1), before = 1, after = 1)
+  })
+  expect_snapshot(error = TRUE, {
+    relocate_loc(c(x = 1), before = 1, after = 1, before_arg = ".before", after_arg = ".after")
+  })
+})
+
+test_that("can't relocate with out-of-bounds variables by default", {
+  x <- c(a = 1, b = 2)
+
+  expect_snapshot({
+    (expect_error(relocate_loc(x, c)))
+    (expect_error(relocate_loc(x, a, before = c)))
+    (expect_error(relocate_loc(x, a, after = c)))
+  })
+})
+
+test_that("can relocate with out-of-bounds variables in `expr` if `strict = FALSE`", {
+  x <- c(a = 1, b = 2)
+
+  expect_identical(relocate_loc(x, c, strict = FALSE), c(a = 1L, b = 2L))
+  expect_identical(relocate_loc(x, c(d = b, e = c), strict = FALSE), c(d = 2L, a = 1L))
+
+  # But still not with OOB variables in `before` or `after`
+  expect_snapshot({
+    (expect_error(relocate_loc(x, a, before = c, strict = FALSE)))
+    (expect_error(relocate_loc(x, a, after = c, strict = FALSE)))
+  })
```

Comment:
I've chosen not to pass `strict` on to the eval-select call for `before` and `after`, but don't have a strong preference if anyone disagrees with this.

If we passed `strict` on and `strict = FALSE` and `before = c` was used here, it would look like an empty `before` selection which moves `a` to the front. For the `after` case it would move `a` to the end.

I mainly didn't feel like we needed to try and support / document this

Reply:
I think you did well, `strict` is more or less in the questioning stage so we don't need to support it fully.